packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx

```diff
@@ -45,46 +51,71 @@ function Breadcrumbs<T>(props: SpectrumBreadcrumbsProps<T>, ref: DOMRef) {
+  let defaultVisibleItems: number;
+
+  if (isCollapsible) {
+    defaultVisibleItems = childArray.length;
+  } else {
+    defaultVisibleItems = showRoot ? maxVisibleItems as number - 1 : maxVisibleItems as number;
+  }
+
+  const [visibleItems, setVisibleItems] = useState(defaultVisibleItems);
 
   let {breadcrumbsProps} = useBreadcrumbs(props);
   let {styleProps} = useStyleProps(otherProps);
 
   useEffect(() => {
-    let listItems = [...listRef.current.children];
-    let childrenWidthTotals = listItems.reduce((acc, item, index) => (
-      [...acc, acc[index] + item.getBoundingClientRect().width]
-    ), [0]);
-
-    let onResize = () => {
-      if (isCollapsible && listRef.current) {
+    // Only run the resize logic if the menu is collapsible to avoid the risk of performance problems.
+    if (isCollapsible && listRef.current) {
+      let listItems = [...listRef.current.children];
+      // Ignore the last item when the size is large because it wraps onto a new line and doesn't take up horizontal space.
+      let listItemsToMeasure = size === 'L' ? listItems.slice(0, listItems.length - 1) : listItems;
+      let childrenWidths = listItemsToMeasure.map((item) => item.getBoundingClientRect().width);
+
+      let onResize = () => {
         let containerWidth = listRef.current.getBoundingClientRect().width;
-        let index = childrenWidthTotals.findIndex(totalWidth => totalWidth > containerWidth);
+        let [rootBreadcrumbWidth, ...otherBreadcrumbWidths] = childrenWidths;
+        let calculatedWidth = 0;
+
+        // Make sure we account for the root breadcrumb if it's enabled.
+        if (showRoot) {
+          calculatedWidth += rootBreadcrumbWidth;
+        }
+
+        let otherVisibleItemsCount = 0;
+
+        // See how many other breadcrumbs we can fit (starting from the right).
+        otherBreadcrumbWidths.reverse().forEach(breadcrumbWidth => {
+          calculatedWidth += breadcrumbWidth;
+          if (calculatedWidth < containerWidth) {
+            otherVisibleItemsCount++;
+          }
+        });
 
-        let visibleItemsCount;
         let minVisibleItems = showRoot ? MIN_VISIBLE_ITEMS + 1 : MIN_VISIBLE_ITEMS;
 
-        if (index ===  -1) {
-          visibleItemsCount = childArray.length;
-        } else if (index < minVisibleItems) {
-          visibleItemsCount = minVisibleItems;
-        } else {
-          visibleItemsCount = index;
+        if (otherVisibleItemsCount < minVisibleItems) {
+          otherVisibleItemsCount = MIN_VISIBLE_ITEMS;
         }
 
-        setVisibleItems(visibleItemsCount);
-      }
-    };
+        let maxVisibleOtherItems = showRoot ? MAX_VISIBLE_ITEMS - 1 : MAX_VISIBLE_ITEMS;
 
-    window.addEventListener('resize', onResize);
-    onResize();
-    return () => {
-      window.removeEventListener('resize', onResize);
-    };
-  }, [isCollapsible, childArray.length, listRef, showRoot]);
+        if (isCollapsible && otherVisibleItemsCount > maxVisibleOtherItems) {
```

Comment:
I don't think you need to check isCollapsible given that this function is defined inside that check and a new useEffect run will go if isCollapsible changes